Cesium3DTilesSelection/src/TileWorkManager.cpp

```diff
@@ -0,0 +1,571 @@
+    return;
+
+  {
+    std::lock_guard<std::mutex> lock(thiz->_requestsLock);
+    thiz->_maxSimultaneousRequests = maxSimultaneousRequests;
+
+    // Copy load requests into internal work we will own
+    thiz->ordersToWork(requestOrders, workCreated);
+    thiz->ordersToWork(processingOrders, workCreated);
+  }
+
+  if (requestOrders.size())
+    transitionRequests(thiz);
+}
+
+void TileWorkManager::RequeueWorkForRequest(
+    std::shared_ptr<TileWorkManager>& thiz,
+    Work* work) {
+  {
+    std::lock_guard<std::mutex> lock(thiz->_requestsLock);
+
+    if (thiz->_shutdownSignaled)
+      return;
+
+    // This processing work should be in flight, remove it
+    auto foundIt = thiz->_processingInFlight.find(work->uniqueId);
+    assert(foundIt != thiz->_processingInFlight.end());
+    thiz->_processingInFlight.erase(foundIt);
+
+    thiz->stageWork(work);
+  }
+
+  transitionRequests(thiz);
+}
+
+void TileWorkManager::SignalWorkComplete(Work* work) {
+  std::lock_guard<std::mutex> lock(_requestsLock);
+
+  if (_shutdownSignaled)
+    return;
+
+  // Assert this work is already owned by this manager
+  assert(_ownedWork.find(work->uniqueId) != _ownedWork.end());
+
+  // This processing work should be in flight, remove it
+  auto foundIt = _processingInFlight.find(work->uniqueId);
+  assert(foundIt != _processingInFlight.end());
+  _processingInFlight.erase(foundIt);
+
+  // Assert this is not in any other queues
+#ifndef NDEBUG
+  for (auto element : _requestsPending)
+    assert(element->uniqueId != work->uniqueId);
+  for (auto& urlWorkVecPair : _requestsInFlight)
+    for (auto element : urlWorkVecPair.second)
+      assert(element->uniqueId != work->uniqueId);
+  for (auto element : _processingPending)
+    assert(element->uniqueId != work->uniqueId);
+#endif
+
+  // Put in the done list
+  _doneWork.push_back(work);
+}
+
+void TileWorkManager::TryDispatchProcessing(
+    std::shared_ptr<TileWorkManager>& thiz) {
+  transitionProcessing(thiz);
+}
+
+void TileWorkManager::onRequestFinished(
+    std::shared_ptr<IAssetRequest>& pCompletedRequest) {
+
+  std::lock_guard<std::mutex> lock(_requestsLock);
+
+  if (_shutdownSignaled)
+    return;
+
+  const IAssetResponse* response = pCompletedRequest->response();
+  uint16_t responseStatusCode = response ? response->statusCode() : 0;
+
+  // Find this request
+  auto foundIt = _requestsInFlight.find(pCompletedRequest->url());
+  assert(foundIt != _requestsInFlight.end());
+
+  // Handle results
+  std::vector<Work*>& requestWorkVec = foundIt->second;
+  for (Work* requestWork : requestWorkVec) {
+    CesiumAsync::RequestData* workNextRequest = requestWork->getNextRequest();
+    assert(workNextRequest);
+    assert(pCompletedRequest->url() == workNextRequest->url);
+
+    assert(_ownedWork.find(requestWork->uniqueId) != _ownedWork.end());
+
+    // A response code of 0 is not a valid HTTP code
+    // and probably indicates a non-network error.
```

Comment:
No, it probably indicates a valid response from a `file:///` URL. It's annoying that libcurl returns a status code of 0 for a successful file read, but it does.